### ISSUE TITLE
BUG with function autodiff()

### ISSUE THREAD

**Issue by frankwswang** [QuantumBFS/Yao.jl#170]
`autodiff()` can't apply to `ChainBlock` when it consists of an empty `ChainBlock` `chain(N)`.

Example Code:
```
julia> using Yao

julia> using QuAlgorithmZoo

julia> genblock = random_diff_circuit(2, 1, pair_ring(2))
Total: 2, DataType: Complex{Float64}
chain
├─ chain
│  ├─ chain
│  │  ├─ put on (1)
│  │  │  └─ Rot X gate: 0.0
│  │  └─ put on (1)
│  │     └─ Rot Z gate: 0.0
│  └─ chain
│     ├─ put on (2)
│     │  └─ Rot X gate: 0.0
│     └─ put on (2)
│        └─ Rot Z gate: 0.0
├─ [↺] chain
│  ├─ control(1)
│  │  └─ (2,)=>X gate
│  └─ control(2)
│     └─ (1,)=>X gate
└─ chain
   ├─ chain
   │  ├─ put on (1)
   │  │  └─ Rot Z gate: 0.0
   │  └─ put on (1)
   │     └─ Rot X gate: 0.0
   └─ chain
      ├─ put on (2)
      │  └─ Rot Z gate: 0.0
      └─ put on (2)
         └─ Rot X gate: 0.0


julia> gen1 = chain(2, genblock, genblock) |> autodiff(:QC)
Total: 2, DataType: Complex{Float64}
chain
├─ chain
│  ├─ chain
│  │  ├─ chain
│  │  │  ├─ put on (1)
│  │  │  │  └─ [̂∂] Rot X gate: 0.0
│  │  │  └─ put on (1)
│  │  │     └─ [̂∂] Rot Z gate: 0.0
│  │  └─ chain
│  │     ├─ put on (2)
│  │     │  └─ [̂∂] Rot X gate: 0.0
│  │     └─ put on (2)
│  │        └─ [̂∂] Rot Z gate: 0.0
│  ├─ [↺] chain
│  │  ├─ control(1)
│  │  │  └─ (2,)=>X gate
│  │  └─ control(2)
│  │     └─ (1,)=>X gate
│  └─ chain
│     ├─ chain
│     │  ├─ put on (1)
│     │  │  └─ [̂∂] Rot Z gate: 0.0
│     │  └─ put on (1)
│     │     └─ [̂∂] Rot X gate: 0.0
│     └─ chain
│        ├─ put on (2)
│        │  └─ [̂∂] Rot Z gate: 0.0
│        └─ put on (2)
│           └─ [̂∂] Rot X gate: 0.0
└─ chain
   ├─ chain
   │  ├─ chain
   │  │  ├─ put on (1)
   │  │  │  └─ [̂∂] Rot X gate: 0.0
   │  │  └─ put on (1)
   │  │     └─ [̂∂] Rot Z gate: 0.0
   │  └─ chain
   │     ├─ put on (2)
   │     │  └─ [̂∂] Rot X gate: 0.0
   │     └─ put on (2)
   │        └─ [̂∂] Rot Z gate: 0.0
   ├─ [↺] chain
   │  ├─ control(1)
   │  │  └─ (2,)=>X gate
   │  └─ control(2)
   │     └─ (1,)=>X gate
   └─ chain
      ├─ chain
      │  ├─ put on (1)
      │  │  └─ [̂∂] Rot Z gate: 0.0
      │  └─ put on (1)
      │     └─ [̂∂] Rot X gate: 0.0
      └─ chain
         ├─ put on (2)
         │  └─ [̂∂] Rot Z gate: 0.0
         └─ put on (2)
            └─ [̂∂] Rot X gate: 0.0


julia> gen2 = chain(2, chain(2), genblock, genblock) |> autodiff(:QC)
ERROR: MethodError: no method matching Yao.Blocks.ChainBlock(::Array{Any,1})
```

Please fix this BUG, thank you very much!

**Comment by GiggleLiu** [comment]
Sorry, can not fix this issue rightnow for `Yao` is under a reconstruction.

To solve the problem, You may put these codes on top of your own project 

```julia
import Yao: autodiff

function autodiff(mode::Val{:QC}, blk::AbstractBlock)
   blks = subblocks(blk)
   isempty(blks) ? blk : chsubblocks(blk, autodiff.(mode, blks))
end
```

Thanks for your feedback!

**Comment by Roger-luo** [comment]
@GiggleLiu actually We can fix this... and add support to v0.3.0. It's in v0.3.2 now.

@frankwswang  but please notice Yao does not have any LTS (long time support) version at the moment (the v1.0 version), but we are approaching it, hopefully v0.4 is going to be the last beta version, thus, you will be asked to update your code when v0.4 is released.

It's fixed in v0.3.2 and as long as it's merged to METADATA (https://github.com/JuliaLang/METADATA.jl/pull/22681), you can fix this by simply updating your Yao package by the following command

```
] up Yao
```

But we won't be able to fix any bug next week since Julia itself is migrating the whole package registration to the registrator, and there might be some glitches for registering new version.
